stat.go

```diff
@@ -648,6 +648,94 @@ func KullbackLeibler(p, q []float64) float64 {
 	return kl
 }
 
+// LinearRegression computes the best-fit line
+//  y = alpha*x + beta
+// to the data in x and y with the given weights. If origin is true, the
+// regression is forced to pass through the origin.
+//
+// Specifically, LinearRegression computes the values of alpha and
+// beta such that the total residual
+//  \sum_i w[i]*(y[i] - alpha*x[i] - beta)^2
+// is minimized. If origin is true, then beta is forced to be zero.
+//
+// The lengths of x and y must be equal. If weights is nil then all of the
+// weights are 1. If weights is not nil, then len(x) must equal len(weights).
+func LinearRegression(x, y, weights []float64, origin bool) (alpha, beta float64) {
+	if len(x) != len(y) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(weights) != len(x) {
+		panic("stat: slice length mismatch")
+	}
+
+	w := 1.0
+	if origin {
+		var x2Sum, xySum float64
+		for i, xi := range x {
+			if weights != nil {
+				w = weights[i]
+			}
+			yi := y[i]
+			xySum += w * xi * yi
+			x2Sum += w * xi * xi
+		}
+		beta = xySum / x2Sum
+
+		return alpha, beta
+	}
+
+	beta = Covariance(x, y, weights) / Variance(x, weights)
+	alpha = Mean(y, weights) - beta*Mean(x, weights)
+
+	return alpha, beta
+}
+
+// RSquared returns the coefficient of determination for the best-fit
+// line
+//  y = alpha*x + beta
+// and the data in x and y with the given weights.
+//
+// The lengths of x and y must be equal. If weights is nil then all of the
+// weights are 1. If weights is not nil, then len(x) must equal len(weights).
+func RSquared(x, y, weights []float64, alpha, beta float64) float64 {
+	if len(x) != len(y) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(weights) != len(x) {
+		panic("stat: slice length mismatch")
+	}
+
+	w := 1.0
+	if alpha == 0 {
+		var ssr, tot float64
+		for i, xi := range x {
+			if weights != nil {
+				w = weights[i]
+			}
+			fi := beta * xi
```

Comment:
This doesn't look right either. Shouldn't the residual be `y[i] - beta * x[i]`, and then at the end be 1 - ssr/tot. If you're trying to use the other formula, it needs to use ymean (which is worse, because then we need two passes over the data).